From e47c0b1e42f7 Mon Sep 17 00:00:00 2001
From: Elad Yifee <eladwf@gmail.com>
Date: Mon, 29 Jul 2024 18:29:54 +0300
Subject: [PATCH] net: ethernet: mtk_eth_soc: use prefetch methods

Utilize kernel prefetch methods to improve packet processing
performance. This has shown to improve performance in high packet
rates, about 5% more packets/sec.

v2: use net_prefetchw for skb->data per Joe Damato; use eth->ip_align
instead of NET_IP_ALIGN.

Note: the non-pagepool net_prefetch() hunk is already provided by
999-9907-2 (net_prefetch for non-pagepool path), which is applied
before this patch in the 6.18 tree; it is therefore dropped here.

Signed-off-by: Elad Yifee <eladwf@gmail.com>
Signed-off-by: SmartWRT <smartwrt@localhost>
---
 drivers/net/ethernet/mediatek/mtk_eth_soc.c | 7 +++++++
 1 file changed, 7 insertions(+)

--- a/drivers/net/ethernet/mediatek/mtk_eth_soc.c
+++ b/drivers/net/ethernet/mediatek/mtk_eth_soc.c
@@ -2469,6 +2469,7 @@ static u32 mtk_xdp_run(struct mtk_eth *eth, struct mtk_rx_ring *ring,
 	if (!prog)
 		goto out;
 
+	net_prefetchw(xdp->data_hard_start);
 	act = bpf_prog_run_xdp(prog, xdp);
 	switch (act) {
 	case XDP_PASS:
@@ -2544,6 +2545,7 @@ static int mtk_poll_rx(struct napi_struct *napi, int budget,
 
 		idx = NEXT_DESP_IDX(ring->calc_idx, ring->dma_size);
 		rxd = ring->dma + RX_DESC_OFS(eth, idx);
+		prefetch(rxd);
 		data = ring->data[idx];
 
 		if (!mtk_rx_get_desc(eth, &trxd, rxd, ring->in_sram))
@@ -2611,6 +2613,7 @@ static int mtk_poll_rx(struct napi_struct *napi, int budget,
 			if (ret != XDP_PASS)
 				goto skip_rx;
 
+			net_prefetch(xdp.data_meta);
 			skb = napi_build_skb(data, PAGE_SIZE);
 			if (unlikely(!skb)) {
 				page_pool_put_full_page(ring->page_pool,
@@ -2619,6 +2622,7 @@ static int mtk_poll_rx(struct napi_struct *napi, int budget,
 				goto skip_rx;
 			}
 
+			net_prefetchw(skb->data);
 			skb_reserve(skb, xdp.data - xdp.data_hard_start);
 			skb_put(skb, xdp.data_end - xdp.data);
 			metasize = xdp.data - xdp.data_meta;
@@ -2660,7 +2664,8 @@ static int mtk_poll_rx(struct napi_struct *napi, int budget,
 				goto skip_rx;
 			}
 
-			skb_reserve(skb, NET_SKB_PAD + NET_IP_ALIGN);
+			net_prefetchw(skb->data);
+			skb_reserve(skb, NET_SKB_PAD + eth->ip_align);
 			skb_put(skb, pktlen);
 		}
 
